Data/ByteString/Builder/RealFloat.hs

```diff
@@ -0,0 +1,179 @@
+-- |
+-- Module      : Data.ByteString.Builder.RealFloat
+-- Copyright   : (c) Lawrence Wu 2021
+-- License     : BSD-style
+-- Maintainer  : lawrencejwu@gmail.com
+--
+-- Floating point formatting for Bytestring.Builder
+--
+-- This module primarily exposes `floatDec` and `doubleDec` which do the
+-- equivalent of converting through `string7 . show`.
+--
+-- Experimentally, it also exposes `formatFloat` and `formatDouble` which
+-- accept the same arguments as `GHC.Float.formatRealFloat`. Currently,
+-- precision is not supported for the general and scientific notation versions
+-- of this API.
+--
+
+module Data.ByteString.Builder.RealFloat
+  ( FloatFormat(..)
+  , floatDec
+  , doubleDec
+  , formatFloat
+  , formatDouble
+  ) where
+
+import Data.ByteString.Builder.Internal (Builder)
+import qualified Data.ByteString.Builder.RealFloat.Internal as R
+import qualified Data.ByteString.Builder.RealFloat.F2S as RF
+import qualified Data.ByteString.Builder.RealFloat.D2S as RD
+import qualified Data.ByteString.Builder.Prim as BP
+import GHC.Float (roundTo)
+import GHC.Word (Word32, Word64)
+import GHC.Show (intToDigit)
+
+-- | Returns a rendered Float. Matches `show` in displaying in fixed or
+-- scientific notation
+{-# INLINABLE floatDec #-}
+floatDec :: Float -> Builder
+floatDec = formatFloat FGeneric Nothing
+
+-- | Returns a rendered Double. Matches `show` in displaying in fixed or
+-- scientific notation
+{-# INLINABLE doubleDec #-}
+doubleDec :: Double -> Builder
+doubleDec = formatDouble FGeneric Nothing
+
+-- | ByteString float-to-string format
+data FloatFormat
+  = FExponent       -- ^ scientific notation
+  | FFixed          -- ^ fixed precision with `Maybe Int` digits after the decimal
+  | FGeneric        -- ^ dispatches to fixed or exponent based on the exponent
+  deriving Show
+
+-- TODO: support precision argument for FGeneric and FExponent
```

Comment:
@la-wu What's the reason why you left this for future work? Is it tricky to implement?

It's a bit of a wart that the `Maybe Int` argument is meaningless in the `FExponent` case. I was wondering whether we ought to hide the wart and improve the forward-compatibility by creating an opaque format type somewhat like this:

```haskell
data FloatFormat = MkFloatFormat
  { floatFormat' :: FloatFormat' -- the old FF type
  , precision :: Maybe Int
  }

fixed :: Int -> FloatFormat
fixed n = MkFloatFormat FFixed (Just n)

scientific :: FloatFormat
scientific = MkFloatFormat FExponent Nothing

-- added once the precision support is complete
scientificWithPrecision :: Int -> FloatFormat
scientific n = MkFloatFormat FExponent (Just n)

-- generic FloatFormats
...
```

Reply:
It was mainly left unimplemented because we previously only had `floatDec` and `doubleDec` so I did not spend too much time on it. I implemented the `FFixed` case to match `show` and what we had before (we need to use non-exponent formatting for `e >= 0 && e <= 7`).

I don't think it would be particularly hard to implement something for precision for FExponent (we should be able to largely follow what is in base https://github.com/ghc/packages-base/blob/master/GHC/Float.lhs#L602 as I did for the fixed-format). But ideally we would move to using a raw buffer-formatter for both FFixed and FExponent with precision so they could be equally efficient there.

Reply:
@Bodigrim How do you feel about future-proofing the API along the lines I suggested above?

An additional advantage of an opaque type is that I think we might get away with not providing any instances for it. With public constructors, I think we ought to provide `Show`, `Ord`, `Enum` etc.

Reply:
I think an opaque type is a good idea.

Reply:
Took a stab at this. Is this what you guys had in mind? https://github.com/haskell/bytestring/pull/365/commits/0474332d4fca7f757e024af5f2b6449b73d04672